File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/StatisticsPacket.java
package gg.mineral.server.network.packet.play.clientbound;

import gg.mineral.api.network.packet.Packet;

import io.netty.buffer.ByteBuf;
import it.unimi.dsi.fastutil.objects.Object2IntMap;
import lombok.val;

public final record StatisticsPacket(Object2IntMap<String> statistics) implements Packet.OUTGOING {

    @Override
    public void serialize(ByteBuf os) {
        writeVarInt(os, statistics.size());

        for (val entry : statistics.object2IntEntrySet()) {
            writeString(os, entry.getKey());
            writeVarInt(os, entry.getIntValue());
        }
    }

    @Override
    public byte getId() {
        return 0x37;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/play/clientbound/StatisticsPacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf
import it.unimi.dsi.fastutil.objects.Object2IntMap

@JvmRecord
data class StatisticsPacket(val statistics: Object2IntMap<String>) : Packet.Outgoing {
    override fun serialize(os: ByteBuf) {
        os.writeVarInt(statistics.size)

        for (entry in statistics.object2IntEntrySet()) {
            os.writeString(entry.key)
            os.writeVarInt(entry.intValue)
        }
    }

    override val id: Byte
        get() = 0x37
}
